File: examples/worker-poller.example.ts
/**
 * Example of what a Node worker client would look like.
 * The real clipper (Python) will do the equivalent using requests / httpx.
 *
 * Usage:
 *   X_MACHINE_TOKEN=mt_xxx npx ts-node examples/worker-poller.example.ts
 */

// NOTE: This is an illustrative example only.
// No external HTTP lib required — uses global fetch (Node 18+ / modern runtimes).

const BASE = process.env.API_BASE || 'http://localhost:3001';
const TOKEN = process.env.X_MACHINE_TOKEN || 'dev_machine_token_please_change';

const HEADERS = {
  'X-Machine-Token': TOKEN,
  'Content-Type': 'application/json',
};

async function apiGet<T>(path: string, params?: Record<string, any>): Promise<T> {
  const url = new URL(path, BASE);
  if (params) Object.entries(params).forEach(([k, v]) => url.searchParams.set(k, String(v)));
  const res = await fetch(url, { headers: HEADERS, signal: AbortSignal.timeout(35_000) });
  if (!res.ok) throw new Error(`GET ${path} failed: ${res.status}`);
  return res.json() as Promise<T>;
}

async function apiPatch<T>(path: string, body: any): Promise<T> {
  const res = await fetch(new URL(path, BASE), {
    method: 'PATCH',
    headers: HEADERS,
    body: JSON.stringify(body),
    signal: AbortSignal.timeout(15_000),
  });
  if (!res.ok) throw new Error(`PATCH ${path} failed: ${res.status}`);
  return res.json() as Promise<T>;
}

interface LeasedJob {
  id: string;
  type: string;
  payload: {
    shareCode: string;
    trustedSteamIds?: string[];
    options?: Record<string, any>;
  };
  attempts: number;
  leaseExpiresAt: string;
}

async function pollLoop() {
  console.log('Worker starting. Polling for clip jobs...');

  while (true) {
    try {
      // Long-poll up to ~25s
      const { job } = await apiGet<{ job: LeasedJob | null }>('/worker/jobs/lease', { wait: 25 });

      if (!job) {
        console.log('No job, sleeping 3s...');
        await new Promise(r => setTimeout(r, 3000));
        continue;
      }

      console.log('Leased job', job.id, job.payload.shareCode);

      // TODO: run the actual clipper here:
      // e.g. spawn python clipper.py fetch ...
      // stream progress via PATCH:

      await report(job.id, { status: 'PROCESSING', stage: 'downloading', progress: 5, message: 'Fetching demo' });

      // ... during render
      await report(job.id, { progress: 35, stage: 'rendering', message: 'Recording clip 2/5' });

      // on finish
      await report(job.id, {
        status: 'COMPLETED',
        progress: 100,
        stage: 'done',
        result: {
          clips: [{ id: 'clip_001_4k', url: 'https://...' }],
          manifestUrl: 'https://...',
        },
      });

      console.log('Job finished', job.id);
    } catch (err: any) {
      console.error('Lease or report error', err?.message || err);
      await new Promise(r => setTimeout(r, 5000));
    }
  }
}

async function report(jobId: string, body: any) {
  await apiPatch(`/worker/jobs/${jobId}`, body);
}

pollLoop().catch(console.error);

File: examples/worker-poller.example.py
"""
Minimal Python worker loop for the clipper backend (illustrative).

Real integration will live inside / next to cs2-clip/clipper.py or a new worker.py.

Run:
    pip install httpx
    X_MACHINE_TOKEN=dev_machine_token_please_change python examples/worker-poller.example.py
"""
import os
import time
import httpx
from typing import Optional, Any

BASE_URL = os.getenv("API_BASE", "http://localhost:3001")
TOKEN = os.getenv("X_MACHINE_TOKEN", "dev_machine_token_please_change")
HEADERS = {"X-Machine-Token": TOKEN}

def lease_job(wait: int = 25) -> Optional[dict]:
    with httpx.Client(timeout=30) as client:
        r = client.get(f"{BASE_URL}/worker/jobs/lease", params={"wait": wait}, headers=HEADERS)
        r.raise_for_status()
        return r.json().get("job")

def report(job_id: str, **kwargs: Any) -> dict:
    with httpx.Client(timeout=15) as client:
        r = client.patch(f"{BASE_URL}/worker/jobs/{job_id}", json=kwargs, headers=HEADERS)
        r.raise_for_status()
        return r.json()

def run_clip_job(job: dict):
    payload = job["payload"]
    share_code = payload["shareCode"]
    trusted = payload.get("trustedSteamIds", [])
    opts = payload.get("options", {})

    print(f"[worker] Starting clip for {share_code} (trusted={trusted}) opts={opts}")

    # === Replace the block below with real calls into clipper ===
    report(job["id"], status="PROCESSING", stage="downloading", progress=3, message="Downloading demo")
    time.sleep(2)
    report(job["id"], progress=18, stage="parsing", message="Parsing demo for clips")
    time.sleep(3)

    # Simulate rendering progress
    for i in range(1, 6):
        report(job["id"], progress=20 + i*12, stage="rendering", message=f"Rendering clip {i}/5")
        time.sleep(1.5)

    report(
        job["id"],
        status="COMPLETED",
        progress=100,
        stage="done",
        result={
            "shareCode": share_code,
            "clips": [
                {"filename": "clip_001_4k.mp4", "url": "https://clips.example/xxx.mp4"}
            ],
        },
    )
    print("[worker] Job completed")

def main():
    print("Clipper worker starting (polling HTTPS)...")
    while True:
        try:
            job = lease_job(wait=20)
            if not job:
                time.sleep(2)
                continue
            run_clip_job(job)
        except Exception as exc:
            print("Worker loop error:", exc)
            time.sleep(5)

if __name__ == "__main__":
    main()
